modin/experimental/xgboost/xgboost.py

```diff
@@ -0,0 +1,499 @@
+):
+
+    row_partitions_by_actors = []
+    if evenly_data_distribution:
+        num_actors = len(actors)
+        row_parts_last_idx = (
+            len(row_partitions) // num_actors
+            if len(row_partitions) % num_actors == 0
+            else len(row_partitions) // num_actors + 1
+        )
+        start_idx = 0
+        for actor in actors:
+            idx_slice = (
+                slice(start_idx, start_idx + row_parts_last_idx)
+                if start_idx + row_parts_last_idx < len(row_partitions)
+                else slice(start_idx, len(row_partitions))
+            )
+            row_partitions_by_actors.append(row_partitions[idx_slice])
+            start_idx += row_parts_last_idx
+    else:
+        actors_ips = [ray.get(actor.get_actor_ip.remote()) for actor in actors]
+        partitions_ips = [ray.get(row[0]) for row in row_partitions]
+
+        # If one of actors isn't contain partitions we make evenly
+        # distribution of data by nodes.
+        if (
+            len(np.unique(actors_ips)) != len(np.unique(partitions_ips))
+            and data_for_aligning is None
+        ):
+            return _get_evenly_distr_of_partitions_by_ips(
+                actors, actors_ips, row_partitions
+            )
+
+        if data_for_aligning is None:
+            for actor_ip in actors_ips:
+                actor_row_partitions = []
+                order_of_indexes = []
+                for i, row in enumerate(row_partitions):
+                    if partitions_ips[i] in actor_ip:
+                        actor_row_partitions.append(row[1])
+                        order_of_indexes.append(i)
+
+                row_partitions_by_actors.append(
+                    (actor_ip, order_of_indexes, actor_row_partitions)
+                )
+        else:
+            for (actor_ip, order_of_indexes, _) in data_for_aligning:
+                actor_row_partitions = []
+                for row_idx in order_of_indexes:
+                    actor_row_partitions.append(row_partitions[row_idx][1])
+
+                row_partitions_by_actors.append(
+                    (actor_ip, order_of_indexes, actor_row_partitions)
+                )
+
+    return row_partitions_by_actors
+
+
+def _get_evenly_distr_of_partitions_by_ips(actors, actors_ips, row_partitions):
+    num_actors = len(actors)
+    row_parts_last_idx = (
+        len(row_partitions) // num_actors
+        if len(row_partitions) % num_actors == 0
+        else len(row_partitions) // num_actors + 1
+    )
+    row_partitions_by_actors = []
+    start_idx = 0
+    for actor, ip in zip(actors, actors_ips):
+        first = start_idx
+        last = (
+            (start_idx + row_parts_last_idx)
+            if (start_idx + row_parts_last_idx < len(row_partitions))
+            else len(row_partitions)
+        )
+
+        actor_row_partitions = []
+        order_of_indexes = []
+        while first < last:
```

Comment:
umm, `for first in range(start_idx, last)`? :)

Reply:
fixed